### ISSUE TITLE
Un/lifting continuations

### ISSUE THREAD

**Issue by brendanhay** [haskell-effectful/effectful#62]
The following contrived example is taken from a larger body of code where I'm attempting to discharge/eliminate effects from a larger stack, in the presence of continuations.

What I want:

```haskell
import Effectful
import Effectful.Concurrent (Concurrent)
import Effectful.Concurrent.Async qualified as Async
import Effectful.Reader.Static (Reader)
import Effectful.Reader.Static qualified as Reader

runDefault :: '[Concurrent] :>> es => (Int -> Eff es ()) -> Eff es ()
runDefault k = Reader.runReader 4 (run k)

run :: '[Concurrent, Reader Int] :>> es => (Int -> Eff es ()) -> Eff es ()
run k = do
  thread <- Reader.ask @Int
  asyncs <- traverse (\n -> Async.async (k n)) [1 .. thread]

  (_exit, ()) <- Async.waitAnyCancel asyncs

  pure ()
```

What I get:

```haskell
runDefault :: '[Concurrent] :>> es => (Int -> Eff (Reader Int ': es) ()) -> Eff es ()
runDefault k = Reader.runReader 4 (run k)

run :: '[Concurrent, Reader Int] :>> es => (Int -> Eff es ()) -> Eff es ()
run k = do
  thread <- Reader.ask @Int
  asyncs <- traverse (\n -> Async.async (k n)) [1 .. thread]

  (_exit, ()) <- Async.waitAnyCancel asyncs

  pure ()
```

Obviously the above isn't going to work as I've not specified any way to un/lift `k` - but I've run into this idiom of having a function such as `run` which exposes all possible effects in use and then a function such as `runDefault` which eliminates `>=1` effects that may just be internal.

Is it possible to un/lift the continuation `k` or use something like `SuffixOf` to state that only a subset of `es` is available to `k`?

**Comment by arybczak** [comment]
Also, what are you trying to do? :thinking:

**Comment by brendanhay** [comment]
> What is `runAsync`?

Typo, it's just `async`.

> Also, what are you trying to do?

Un/lift the supplied `(Int -> Eff es ())` continuation such that it has a "smaller" environment `es` than the parent? environment - something like `Control.Lens.zoom` - in the example above I use `Reader Int` in the parent but this should not be available to the `es` in `(Int -> Eff es ())` - I lack the language to explain it any better than that if the example is non-obvious.

**Comment by brendanhay** [comment]
(`runDefault` above would be forced to expose `Reader Int` despite having used `runReader` in an attempt to discharge the effect.)

**Comment by brendanhay** [comment]
I updated the example in an attempt to be clearer.

**Comment by arybczak** [comment]
IIUC, you can do this:

```haskell
import Effectful
import Effectful.Concurrent
import Effectful.Concurrent.Async
import Effectful.Reader.Static

runDefault
  :: '[Concurrent] :>> es
  => (forall esK. [Concurrent, Reader Int] :>> esK => Int -> Eff esK ())
   -> Eff es ()
runDefault k = runReader @Int 4 $ run k

run :: '[Concurrent, Reader Int] :>> es => (Int -> Eff es ()) -> Eff es ()
run k = do
  thread <- ask @Int
  asyncs <- traverse (\n -> async (k n)) [1 .. thread]

  (_exit, ()) <- waitAnyCancel asyncs

  pure ()
```

**Comment by arybczak** [comment]
Does that do what you want?

**Comment by brendanhay** [comment]
Sorry, no - I've been trying to think about how to better enunciate what's going on.

Basically by using a `forall esK.` you're stating that the continuation can only use those two effects - where I'm trying to prevent the fact `runReader` is used from "leaking" to the caller. `localSeqUnlift` have something that looks like the right type signature for unlifting the continuation to have a "smaller" `esK`, but require you produce a `LocalEnv`.

**Comment by arybczak** [comment]
Right, so something similar to `localSeqUnlift` currently cannot be written in such context. It's relatively easy to make it possible, but before I do that I'd like to better understand how did you end up with such construction :thinking: 

Because it looks to me that this wouldn't work with mtl-style effects either. Are you converting existing code that used e.g. ReaderT pattern to `effectful`?

**Comment by arybczak** [comment]
Checkout the `add-raise` branch.

Now this works:

```haskell
import Effectful
import Effectful.Concurrent
import Effectful.Concurrent.Async
import Effectful.Reader.Static

runDefault :: '[Concurrent] :>> es => (Int -> Eff es ()) -> Eff es ()
runDefault k = runReader @Int 4 (run $ raise . k)

run :: '[Concurrent, Reader Int] :>> es => (Int -> Eff es ()) -> Eff es ()
run k = do
  thread <- ask @Int
  asyncs <- traverse (\n -> async (k n)) [1 .. thread]

  (_exit, ()) <- waitAnyCancel asyncs

  pure ()
```

This is what you want, right?

I'll add this to `master`, just need to make the implementation a bit more robust wrt. error handling.

**Comment by brendanhay** [comment]
> This is what you want, right?

That's correct - looks perfect!

> Because it looks to me that this wouldn't work with mtl-style effects either. Are you converting existing code that used e.g. ReaderT pattern to effectful?

No, this is new code that is fairly heavy on the user supplying continuations into a framework for customisation/callbacks, where all possible effects from within the framework seemed to be forced to bubble up to the user which seemed undesirable as it exposes the internal implementation of said framework. Probably I would've designed this different under MTL but this seemed like the most natural/direct approach, so I was interested to see if it was possible without having to go down another route, such as using an IORef or something similar internally instead of the `Reader` or `State` effects.

**Comment by arybczak** [comment]
Fixed in 7f4d55595be3a59fbbb335721204f2a098827fc5.

**Comment by brendanhay** [comment]
Thank you very much!

**Comment by arybczak** [comment]
FYI current implementation is slightly broken, I need to fix it.

**Comment by brendanhay** [comment]
I meant to raise (ha) an issue about the env invariant check that was
throwing - I assume this is what you’re referring to?

**Comment by arybczak** [comment]
Yes. It should be fixed in 3890b0a600435567673543bc6430cee443026bea.
